### ISSUE TITLE
scatter matmul?

### ISSUE THREAD

**Issue by Wang-Tianyu** [rusty1s/pytorch_scatter#144]
Hi, I recently encountered the problem of implementing point cloud based dot product attention, where a scatter matmul operation is needed.

The input is a matrix of shape ((n1 + n2 + n3 + ... + nb), K) where b is the batch size and ni is the number of points in each point cloud; and a matrix of shape (b, C, K).

Ideally, scatter matmul operation gathers ni (1, K) vectors in each batch to form a matrix of shape (ni, K).
Then a standard matrix multiplication between the (C, K) matrix and (K, ni) matrix is carried out and return a (C, ni) matrix for each batch.
In the end, a ((n1 + n2 + n3 + ... + nb), C) matrix is returned.
Then attention can be easily obtained via a softmax.

Since this operation is essentially a permutation invariant aggregation, I feel like it might be an interesting operation to add to this package.

**Comment by rusty1s** [comment]
I fully agree that this is an interesting and important operation. However, I feel that its implementation is quite tricky since it requires a re-implementation of a more general CUDA matmul.

**Comment by PhilippThoelke** [comment]
Hi, is there any update on this?

@Wang-Tianyu did you find an efficient implementation without requiring a custom CUDA kernel?

**Comment by rusty1s** [comment]
I think you only have two options without a dedicated `scatter_matmul` implementation:
1. Split the points per example and apply a matmul sequentially
2. Try to reshape your matrix into the form `[batch_size, max_num_points, K]` and make use of `einsum` functionality.

**Comment by PhilippThoelke** [comment]
I tried 1., which was just incredibly slow and not feasible in my case. I will try 2. as it is likely faster than my current temporary solution, although I wanted to avoid multiplying lots of zeros. Thank you!
